### ISSUE TITLE
TraP freezing on _flag_many_to_many_tempruncat

### ISSUE THREAD

**Issue by AntoniaR** [transientskp/tkp#432]
If you open the Banana page for a given database when TraP is running, most of the time it works completely fine. However, if you load the Banana page for a particular database at a critical point when TraP is conducting source association it will cause TraP to freeze and completely stop processing the job. If you leave TraP for hours (I have tried >12 hours), it still does not progress so you have to kill the job. This also usually leads to a problematic database that cannot be deleted due to remaining open connections.

This only happens when processing a large dataset through TraP, with a thousand or more runcat sources and trying to access the identical database via Banana at a very critical time. Most of the time it's fine (and never happened with smaller datasets) but with the big datasets I think you're more likely to find that critical point because the source association takes longer.

Is there a way to lock the databases when TraP is using them, at key points like source association, which could prevent Banana from reading them at that time?

**Comment by jdswinbank** [comment]
There's a fundamental issue here: reading from the database should not impact on ongoing processing. The database itself should be capable of multiplexing queries without deadlocking.

Whacking some ugly great lock in as part of the pipeline might be a quick & dirty workaround, but it sounds like just papering over a more fundamental problem.

**Comment by timstaley** [comment]
Interesting. @AntoniaR, is with a Postgres or Monetdb database?

**Comment by AntoniaR** [comment]
Postgres. These are images with >1000 sources per image, so source assoc can be slow.

**Comment by timstaley** [comment]
OK. Hoping @gijzelaerr can tell us what the postgres config is like on the UvA machine. On our machines I just gave Banana a read/write login, because it was simplest to set up, but it's possible to configure postgres with read-only logins, which might help. 

Beyond that, as John points out this sort of isn't our responsibility (which is not to say it isn't a problem), it should be up to  Postgres to handle high concurrent loads (and I'm impressed that you managed to break it!). Unfortunately this sounds like the sort of bug that could be hard to reproduce, which would be a requirement for us to submit a useful upstream bug report to the Postgres people. Maybe we can try testing this once we've got a benchmarking suite set up to stress-test the database, but I'd suggest that unless we can identify a clear cause, we don't hold up release 2.1 for this issue. Does that sound reasonable, or is this a complete show-stopper for you?

**Comment by timstaley** [comment]
One more thought - if reworking the query system as Gijs has mentioned goes well, that might help issues, which is another reason to kick this down the road a bit.

**Comment by gijzelaerr** [comment]
Ok it looks like something went wrong doing an `INSERT` or `UPDATE` in `AR_MWA_trans2` and `AR_MWA_trans`. This is *or* a bug in TraP where some query is going wild and gets in an infinite loop *or* this is PostgreSQL that can't handle the query properly. I'm not sure yet what the query is, will try to find out. Currently these processes block read and write access to the tables so guaranty data integrity. 

You can see this here:
```
postgres=# SELECT datname,usename,client_addr,waiting,query_start FROM pg_stat_activity ;
    datname    | usename  | client_addr  | waiting |          query_start          
---------------+----------+--------------+---------+-------------------------------
 AR_MWA_trans2 | antoniar | 146.50.56.73 | f       | 2015-06-17 11:03:53.389381+02
 AR_MWA_trans  | antoniar | 146.50.56.73 | f       | 2015-06-18 06:28:26.938972+02
 AR_MWA_trans2 | antoniar | 146.50.56.73 | f       | 2015-06-19 13:50:02.191424+02
 AR_MWA_trans2 | postgres |              | f       | 2015-06-18 05:23:03.48937+02
 postgres      | postgres |              | f       | 2015-06-19 13:50:02.070086+02
 AR_MWA_trans  | banana   | 127.0.0.1    | t       | 2015-06-19 13:46:22.360723+02
 AR_MWA_trans  | banana   | 127.0.0.1    | t       | 2015-06-18 07:28:47.134039+02
 AR_MWA_trans  | antoniar | 146.50.56.73 | t       | 2015-06-18 07:28:02.003535+02
 AR_MWA_trans  | banana   | 127.0.0.1    | t       | 2015-06-19 09:42:10.822854+02
 AR_MWA_trans  | banana   | 127.0.0.1    | t       | 2015-06-18 07:35:18.949065+02
(10 rows)
```
There are 4 queries running (number 5 is the stat query) which have locked the two databases. Banana and antonia are trying to obtain a lock to the DB, but they are queued since the lock is not released yet. The waiting columns indicate the queued queries.

**Comment by gijzelaerr** [comment]
`AR_MWA_trans2` and `AR_MWA_trans` both are going 100% on the query below, first one started 2 days ago, second query 1 day ago:
```
UPDATE temprunningcatalog
SET inactive = TRUE
WHERE EXISTS (SELECT runcat
,xtrsrc
FROM (SELECT t1.runcat
,t1.xtrsrc
FROM (SELECT xtrsrc
,MIN(r) as min_r
FROM temprunningcatalog
WHERE runcat IN (SELECT runcat
FROM temprunningcatalog
WHERE runcat IN (SELECT runcat
FROM temprunningcatalog
WHERE xtrsrc IN (SELECT xtrsrc
FROM temprunningcatalog
GROUP BY xtrsrc
```

Next to that `AR_MWA_trans2` also has a query going 100%:
```
SELECT id
,ra
,decl
FROM monitor
WHERE dataset = 3
```
4th query also runs on `AR_MWA_trans2` has close to no CPU usage and looks like an administrative query:
```
autovacuum: VACUUM ANALYZE public.temprunningcatalog
```

**Comment by gijzelaerr** [comment]
one can stop a query by running
```
SELECT pg_cancel_backend(32002);
```
Where 32002 is the PID of the postgresql executing the query. I've done this for the `AR_MWA_trans` database to see what would happen. Trying to rerun the query above manually made me realize this is just a small chunk of the actual query

**Comment by gijzelaerr** [comment]
unfortunately the query length stored in the executing log is hard coded truncated to 256 characters.

**Comment by gijzelaerr** [comment]
this is the query where both DB's have problems with: https://github.com/transientskp/tkp/blob/master/tkp/db/associations.py#L791

**Comment by gijzelaerr** [comment]
after killing the `AR_MWA_trans` query banana loads again, and all other queued queries have been processed.

**Comment by timstaley** [comment]
Thanks Gijs - so, problem solved (in the short term)? Sounds like this was not as much a catastrophic crash as I feared. I guess in the longer run we may have to think about how to limit the number of concurrent queries - either by tweaking connection limits for postgres, or queueing Banana requests.

**Comment by gijzelaerr** [comment]
Well, we now know how to examine a PostgreSQL database, to see which queries are being run and get some stats about them. We still don't know why the `_flag_many_to_many_tempruncat` query takes so long (or gets in an infinite loop?) in some cases.

I think the update queries lock the related tables, and when for example banana tries to do a query it is queued untill the update is finished, which takes way too long.

So no, this issue is not yet resolved, but we need to do more examinationand research  to come with a proper plan.

**Comment by AntoniaR** [comment]
Interesting, although I can't really add much. I guess this is probably the first time that the many-to-many queries are really getting taxed. Before, the very small pre-selection region meant that virtually none of the sources were actually getting into many-to-many source associations. I will be monitoring the database connections as I run TraP over the next few weeks and see if I can spot anything odd.

**Comment by AntoniaR** [comment]
It has happened again... This time, I was deliberately not using Banana, there were no other TraP jobs running on Struis at the time it failed and I was not using one of my scripts to query the database. So it doesn't seem to be an issue with TraP and Banana arguing this time... It also only happened a few images into the run, which is the earliest I have seen to date - odd if it is just sheer size of a dataset, unless the queries somehow know about the size of the whole database... I think the only open connection to this database should therefore be from TraP, which has frozen but I have not killed yet... I was not sure if it would be helpful to have everything left as it is.

The affected database is AR_MWA_trans2 - so maybe this time it is just an ongoing problem with that specific database and I should have used a completely new clean database. I will try to set the identical job running with a new database (see if the error is reproducible...) once you've checked if there is anything useful to learn from this database and we've closed the connections (which killing the job doesn't seem to do). I currently have one job running with the AR_MWA_varib database but that should complete in 5-6 hours - so before work time in Amsterdam... I will be online tonight if you want me to do anything - maybe we could get the test with a new clean database running when we're both watching for anything odd? I could send you the link to the trap.log file so you could also watch the progress if that would be helpful.

From the monitoring page of PostgreSQL locks by day, you can see that there is an underlying constant level in AccessShareLock and RowExclusiveLock since Sunday ~8am (corresponding to about the time the job failed). The spikes on top of this are TraP running perfectly normally with a different database (AR_MWA_varib) and I think there are a couple of spikes from my scripts that were querying AR_MWA_varib after jobs were complete. So they all seem to be running fine.

**Comment by gijzelaerr** [comment]
OK good we are starting to get closer to an isolated test case. I really
start to think it is a problem with the query that somehow ends up in an
infinite loop somehow or maybe an integer overflow somewhere.

Let's wait what the run in a new database gives. If it stops we have a
start of an isolated test case, if not we have to examine AR_MWA_trans2 a
bit more to see what goes wrong.
Op 22 jun. 2015 3:51 AM schreef "Antonia Rowlinson" <
notifications@github.com>:

**Comment by AntoniaR** [comment]
Great - would you be able to close the TraP connections to that database? I think that might be the cleanest way to stop that TraP run and clear the connections. I'll then make a new database and try again... There are no other TraP jobs running on Struis at the moment (well, that I can see on 'top').

**Comment by gijzelaerr** [comment]
We dont need to stop that query unless you want to access that specific
database.

Op ma 22 jun. 2015 10:31 schreef Antonia Rowlinson <notifications@github.com

**Comment by AntoniaR** [comment]
I have started up a new database "AR_MWA_trans3" and just set it running... Will let you know when it has passed the previous crash point. See HipChat for the location of the trap.log file.

**Comment by gijzelaerr** [comment]
my theory until now. The problematic query actually only touches `temprunningcatalog`. `temprunningcatalog` is updated *a lot*, many many insert, updates and deletes.
postgresql tables by design need to be vacuumed from time to time to remove dead tuples. This should especially happen for often updated tables. PostgreSQL does a vacuum automatically, based on stats it gathers from table usage:

http://www.postgresql.org/docs/9.3/static/routine-vacuuming.html

The way we use `temprunningcatalog` is different from how tables are often used, we do a lot of inserts and total clearing on the table. It could be the parameters for the auto vacuum are just not correct for this kind of usage. 

somehow at a certain point the temprunningcatalog table is so polluted with dead tuples that because of the exponential growth in running time of the `_flag_many_to_many_tempruncat()` query it becomes way too slow. That happens quite quickly, it grows from running time in couple of ms to a runtime of a couple of days with 'only' about 4000 dead rows. While this long query is running postgresql does a autovacuum of the table  (not sure why yet) the vacuum can't run during the _flag_many_to_many_tempruncat() query, so it can't improve the performance.

the VACUUM command issues an ACCESS EXCLUSIVE lock on a database or table which prevents banana from accessing the database. the banana query gets queued until the other queries are finished, which takes multiple days. Actually banana is just working fine, it is just the landing page that fails to load since it queries all databases.

So for now my idea is to vacuum more (good habit anyway) and see if that works.

I've set these properties to `temprunningcatalog`:
```
ALTER TABLE table_name SET (autovacuum_vacuum_scale_factor = 0.0); 
ALTER TABLE table_name SET (autovacuum_vacuum_threshold = 1000); 
ALTER TABLE table_name SET (autovacuum_analyze_scale_factor = 0.0); 
ALTER TABLE table_name SET (autovacuum_analyze_threshold = 1000);
````

which would trigger a vacuum on every 1000th modification of the table. You can examine the last time a table has been vacuumed with this query:

```
SELECT relname, last_vacuum, last_autovacuum, last_analyze, last_autoanalyze  
FROM pg_stat_all_tables  WHERE schemaname = 'public'; 
```
@AntoniaR can you run an other test run with your data on the AR_MWA_trans2 database?

**Comment by AntoniaR** [comment]
This sounds like a great theory, hopefully it will solve the issues! I have just started the identical run with AR_MWA_trans2 and will let you know how it goes.

As you probably expected, the previous job had finished. If it is helpful (I expect not as you've already identified the query), the error I saw on screen following your cancellation was:
```
ERROR:tkp.db:Query failed: canceling statement due to user request
. Query: UPDATE temprunningcatalog
   SET inactive = TRUE
 WHERE EXISTS (SELECT runcat
                     ,xtrsrc
                 FROM (SELECT t1.runcat
                             ,t1.xtrsrc
                         FROM (SELECT xtrsrc
                                     ,MIN(r) as min_r
                                 FROM temprunningcatalog
                                WHERE runcat IN (SELECT runcat
                                                   FROM temprunningcatalog
                                                  WHERE runcat IN (SELECT runcat
                                                                     FROM temprunningcatalog
                                                                    WHERE xtrsrc IN (SELECT xtrsrc
                                                                                       FROM temprunningcatalog
                                                                                     GROUP BY xtrsrc
                                                                                     HAVING COUNT(*) > 1
                                                                                    )
                                                                  )
                                                 GROUP BY runcat
                                                 HAVING COUNT(*) > 1
                                                )  
                                  AND xtrsrc IN (SELECT xtrsrc
                                                   FROM temprunningcatalog
                                                 GROUP BY xtrsrc
                                                 HAVING COUNT(*) > 1
                                                )  
                               GROUP BY xtrsrc
                              ) t0
                             ,(SELECT runcat
                                     ,xtrsrc
                                     ,r
                                 FROM temprunningcatalog
                                WHERE runcat IN (SELECT runcat
                                                   FROM temprunningcatalog
                                                  WHERE runcat IN (SELECT runcat
                                                                     FROM temprunningcatalog
                                                                    WHERE xtrsrc IN (SELECT xtrsrc
                                                                                       FROM temprunningcatalog
                                                                                     GROUP BY xtrsrc
                                                                                     HAVING COUNT(*) > 1
                                                                                    )
                                                                  )
                                                 GROUP BY runcat
                                                 HAVING COUNT(*) > 1
                                                )  
                                  AND xtrsrc IN (SELECT xtrsrc
                                                   FROM temprunningcatalog
                                                 GROUP BY xtrsrc
                                                 HAVING COUNT(*) > 1
                                                )  
                              ) t1
                        WHERE t0.xtrsrc = t1.xtrsrc
                          AND t0.min_r < t1.r
                      ) t2
                WHERE t2.runcat = temprunningcatalog.runcat
                  AND t2.xtrsrc = temprunningcatalog.xtrsrc
              )
.
Traceback (most recent call last):
  File "/soft/trap/prob421/bin/trap-manage.py", line 10, in <module>
    tkp.management.main()
  File "/soft/trap/prob421/lib/python2.7/site-packages/tkp/management.py", line 360, in main
    args.func(args)
  File "/soft/trap/prob421/lib/python2.7/site-packages/tkp/management.py", line 229, in run_job
    run(args.name, monitor_coords)
  File "/soft/trap/prob421/lib/python2.7/site-packages/tkp/main.py", line 163, in run
    new_source_sigma_margin=new_src_sigma)
  File "/soft/trap/prob421/lib/python2.7/site-packages/tkp/db/associations.py", line 42, in associate_extracted_sources
    _flag_many_to_many_tempruncat()
  File "/soft/trap/prob421/lib/python2.7/site-packages/tkp/db/associations.py", line 851, in _flag_many_to_many_tempruncat
    tkp.db.execute(query, commit=True)
  File "/soft/trap/prob421/lib/python2.7/site-packages/tkp/db/__init__.py", line 23, in execute
    cursor.execute(query, sanitize_db_inputs(parameters))
psycopg2.extensions.QueryCanceledError: canceling statement due to user request
```

The identical job that was running with AR_MWA_trans3 (a clean database) failed in the same way last night when processing image 261. According to the trap.log, it failed at 2015-06-22 21:43:26.133.

**Comment by AntoniaR** [comment]
So it looks like both AR_MWA_trans2 and AR_MWA_trans3 are now stuck. AR_MWA_trans2 appears to have got stuck about 40min ago on image 167 (so different to before and different to AR_MWA_trans3). I will check back in a couple of hours in case it is a temporary glitch.

Bart - as I really need to get ~10,000 images processed before a conference in 2 weeks time I think it might be time to try monetdb again now rather than October... Do you think it will have the same issues with this query?

**Comment by AntoniaR** [comment]
Just had one thought... I'm using release 2.0 and haven't done an init, I just set it going in the same screen as I assumed the changes were only in postgres - is that ok or should I re init TraP and try again?

**Comment by timstaley** [comment]
We haven't made any significant changes to the database code since 2.0, so I don't think that matters.

**Comment by AntoniaR** [comment]
Thanks Tim :) I thought that was the case but decided it was worth checking.

**Comment by gijzelaerr** [comment]
# current status report

this morning there are 2 long running postgres processes running 100%, PID 13406 (AR_MWA_trans2) and 4464 (AR_MWA_trans3). Both are the `_flag_many_to_many_tempruncat()` query. Both DB's also have a 0% cpu autovacuum query running (trans3: pid, `2649` started `06-22 21:43:59` & trans2: pid `11546`, started `06-23 07:38:31`).



# Auto vacuum metadata

## AR_MWA_trans2
```
       relname       |          last_vacuum          |        last_autovacuum        |         last_analyze          |       last_autoanalyze        
---------------------+-------------------------------+-------------------------------+-------------------------------+-------------------------------
 temprunningcatalog  | 2015-06-22 13:48:54.009078+02 | 2015-06-23 07:37:25.627546+02 | 2015-06-22 13:48:57.448655+02 | 2015-06-23 07:37:25.984857+02
```

```
AR_MWA_trans3=# \connect AR_MWA_trans2
You are now connected to database "AR_MWA_trans2" as user "postgres".
AR_MWA_trans2=# analyze verbose temprunningcatalog;
INFO:  analyzing "public.temprunningcatalog"
INFO:  "temprunningcatalog": scanned 1062 of 1062 pages, containing 1883 live rows and 4135 dead rows; 1883 rows in sample, 1883 estimated total rows
```

## AR_MWA_trans3

```
       relname       | last_vacuum |        last_autovacuum        | last_analyze |       last_autoanalyze        
---------------------+-------------+-------------------------------+--------------+-------------------------------
 temprunningcatalog  |             | 2015-06-22 21:43:02.720293+02 |              | 2015-06-22 21:43:02.744028+02
```

```
AR_MWA_trans3=# analyze verbose temprunningcatalog;
INFO:  analyzing "public.temprunningcatalog"
INFO:  "temprunningcatalog": scanned 226 of 226 pages, containing 1650 live rows and 4741 dead rows; 1650 rows in sample, 1650 estimated total rows
ANALYZE
```

**Comment by gijzelaerr** [comment]
It looks like the autovacuum was triggered, but too late. I'm killing the queries, lowering the threshold for `AR_MWA_trans2`, leaving `AR_MWA_trans3` untouched. Please retry.

**Comment by gijzelaerr** [comment]
This PR should fix the issue: https://github.com/transientskp/tkp/pull/436

**Comment by gijzelaerr** [comment]
Antonia has verified that this PR fixes the issue. With that comes a free speedup, how much is to be determined.

**Comment by AntoniaR** [comment]
Yes, it works beautifully, the speed up is fantastic and the test job has now fully completed with no problems.

Trap is noticeably faster to me. It's going to be difficult to quantify the speed up of a full dataset as I never got a job to complete with this dataset, but typically I would expect it to take 1-2 days and this job was ~22 hours.

I can quantify on an image-by-image basis. It used to take ~110 seconds to process a typical image from this problematic dataset and it now takes ~70 seconds per image. i.e. a reduction of ~35%.

**Comment by timstaley** [comment]
:+1:!

**Comment by AntoniaR** [comment]
So I have now processed several thousand images through this fix almost continuously for the past 5 days... It is working beautifully and is good to merge in my opinion. However, I spotted that the fix has failed some tests? #436?

**Comment by gijzelaerr** [comment]
This is actually travis failing, these two need to be accepted

https://github.com/transientskp/tkp/pull/437
https://github.com/transientskp/tkp/pull/434

**Comment by timstaley** [comment]
Closed by #441
